core/src/cell.rs

```diff
@@ -168,65 +136,69 @@ impl<'a> CellProvider for BlockCellProvider<'a> {
     transaction: &'a Transaction,
     seen_inputs: &mut FnvHashSet<OutPoint>,
     cell_provider: &CP,
-) -> ResolvedTransaction<'a> {
-    let input_cells = transaction
-        .input_pts()
-        .iter()
-        .map(|input| {
-            if seen_inputs.insert(input.clone()) {
-                cell_provider.get_cell_status(input)
+) -> Result<ResolvedTransaction<'a>, UnresolvableError> {
+    let (mut dead_out_points, mut unknown_out_points, mut input_cells, mut dep_cells) = (
+        Vec::new(),
+        Vec::new(),
+        Vec::with_capacity(transaction.inputs().len()),
+        Vec::with_capacity(transaction.deps().len()),
+    );
+
+    // skip resolve input of cellbase
+    if !transaction.is_cellbase() {
+        transaction.input_pts().iter().for_each(|out_point| {
+            let cell_status = if seen_inputs.insert(out_point.clone()) {
+                cell_provider.cell(out_point)
             } else {
                 CellStatus::Dead
-            }
-        })
-        .collect();
+            };
 
-    let dep_cells = transaction
-        .dep_pts()
-        .iter()
-        .map(|dep| {
-            if seen_inputs.contains(dep) {
-                CellStatus::Dead
-            } else {
-                cell_provider.get_cell_status(dep)
+            match cell_status {
+                CellStatus::Dead => dead_out_points.push(out_point.clone()),
+                CellStatus::Unknown => unknown_out_points.push(out_point.clone()),
+                CellStatus::Live(cell_meta) => input_cells.push(cell_meta),
             }
         })
-        .collect();
-
-    ResolvedTransaction {
-        transaction,
-        input_cells,
-        dep_cells,
     }
-}
 
-impl<'a> ResolvedTransaction<'a> {
-    pub fn cells_iter(&self) -> Chain<slice::Iter<CellStatus>, slice::Iter<CellStatus>> {
-        self.dep_cells.iter().chain(&self.input_cells)
-    }
-
-    pub fn cells_iter_mut(
-        &mut self,
-    ) -> Chain<slice::IterMut<CellStatus>, slice::IterMut<CellStatus>> {
-        self.dep_cells.iter_mut().chain(&mut self.input_cells)
-    }
-
-    pub fn is_double_spend(&self) -> bool {
-        self.cells_iter().any(CellStatus::is_dead)
-    }
-
-    pub fn is_orphan(&self) -> bool {
-        self.cells_iter().any(CellStatus::is_unknown)
-    }
+    transaction.dep_pts().iter().for_each(|out_point| {
+        let cell_status = if seen_inputs.contains(out_point) {
+            CellStatus::Dead
+        } else {
+            cell_provider.cell(out_point)
+        };
 
-    pub fn is_fully_resolved(&self) -> bool {
-        self.cells_iter().all(CellStatus::is_live)
+        match cell_status {
+            CellStatus::Dead => dead_out_points.push(out_point.clone()),
+            CellStatus::Unknown => unknown_out_points.push(out_point.clone()),
+            CellStatus::Live(cell_meta) => dep_cells.push(cell_meta),
+        }
+    });
+
+    if !dead_out_points.is_empty() {
+        Err(UnresolvableError::Dead(dead_out_points))
```

Comment:
Do we really need to save all failed resolved out_points? why not return an error on the first failure?

What happened if we have both Dead and Unknown resolves?

Reply:
We can early return `Dead` as a short circuit error, but cannot do for `Unknown`, which is used in the tx pool:
https://github.com/nervosnetwork/ckb/blob/cc16678426b76309967f4c8880a4fc9368434741/shared/src/chain_state.rs#L242

Reply:
`Dead` is a higher priority error than `Unknown`, however, to make code easy to understand and debug, I choose to return all error `OutPoint` instead of first one.

Reply:
IMO short-cut return Dead error is more efficient and clear.

Reply:
I think it's better to return error asap in `resolve_transaction`.